### PR TITLE
C2S/SASL

### PR DESCRIPTION
**Pull request by NelsonVides**
An initial refactoring of SASL out of C2S.

I chose accumulators as the SASL state because they also enable tracing trough refs and timing through their timestamps.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/187753) / elasticsearch_and_cassandra_mnesia / cf71561d991343016883f82e77bb70cfae5558ee
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4101/187753/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4101/187753/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%407b5d359d3f9c.2023-08-15_06.51.01/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/188557) / elasticsearch_and_cassandra_mnesia / 7299e6de569000f989860819c9d15e6518e6b8d0
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4101/188557/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4101/188557/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40930fd48667fa.2023-08-21_07.24.49/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/188814) / elasticsearch_and_cassandra_mnesia / 421457fd81354465cd39e060ab8cf7c38dc2ba74
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4101/188814/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4101/188814/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%400e31cf208ab1.2023-08-23_17.59.34/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

File: big_tests/tests/mim_c2s_SUITE.erl [hunk 1]
```diff
@@ -119,6 +121,16 @@ message_sent_to_malformed_jid_results_in_error(Config) ->
         escalus_assert:is_chat_message(<<"Hi!">>, escalus_client:wait_for_stanza(Bob))
     end).
 
+invalid_resource_fails_to_log(Config) ->
```
Comment (chrzaszcz) [inline big_tests/tests/mim_c2s_SUITE.erl:124]: Wasn't this already done in a merged PR? Maybe rebase?

**Review by chrzaszcz — COMMENTED** [review]
I added minor comments.

File: big_tests/tests/mim_c2s_SUITE.erl (hunk 1 above)
Comment (NelsonVides) [reply big_tests/tests/mim_c2s_SUITE.erl]: I think I did and yet dunno why this shows up here, just rebased again.

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/188958) / elasticsearch_and_cassandra_mnesia / abb89dab00676b3482eaf52495b765f7fa8f4b10
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4101/188958/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4101/188958/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40514e3093ea19.2023-08-24_10.24.21/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Review by chrzaszcz — APPROVED** [review]
Looks good :ok_hand:

**Pull request merged**